Update quest lines, Refactor quests

RedSparr0w opened the pull request:
Refactored quests to `App.game.quests`, no longer saved in `player`.
Moved the different quest types into their own folder.
QuestLines has been reworked a little bit.
![image](https://user-images.githubusercontent.com/7288322/89709332-f7aea980-d9d2-11ea-82af-6efcdf7cfc20.png)
Added a small (i) icon that shows the quest line description when hovered.
![image](https://user-images.githubusercontent.com/7288322/89709323-e1085280-d9d2-11ea-8887-da4ebc1fab32.png)
Added ability for custom quest rewards (only for custom quests).
Added ability for gym reward functions (currently only used to start Deoxys quest line)
Deoxys no longer roaming, awarded at the end of the quest.

All previous save data should convert over automatically
  - Current active quests
  - Current quest progress
  - Completed quest
  - Quest xp/level

RedSparr0w committed:
Allow quests to have an initial starting value

RedSparr0w committed:
minor fixes to tutorial quest

RedSparr0w committed:
remove debugger :sweat_smile:

RedSparr0w committed:
create initial base test for a "deoxys" quest line

RedSparr0w committed:
Adjust Quest display

RedSparr0w committed:
Adjust quest line title display

RedSparr0w committed:
Adjust tutorial quest display

RedSparr0w committed:
Update display for tutorial quest p2

RedSparr0w committed:
Move quest types to own folder

RedSparr0w committed:
..

RedSparr0w committed:
Move questHelper to quests, no longer making use of player object

RedSparr0w committed:
Doesn't need to be static

RedSparr0w committed:
reset refreshes

RedSparr0w committed:
Make generateQuestList mroe generic

RedSparr0w committed:
Add 3 random quest

RedSparr0w committed:
Add custom quest rewards

RedSparr0w committed:
Update QuestLineHelper.ts

RedSparr0w committed:
Merge branch 'develop' into questLines

RedSparr0w committed:
Remove Deoxys from roamers

RedSparr0w committed:
Add Deoxys as reward (temp)

RedSparr0w committed:
save QuestLines

RedSparr0w committed:
idk

RedSparr0w committed:
Update game loading

RedSparr0w committed:
Merge branch 'develop' into questLines

RedSparr0w committed:
Merge branch 'develop' into questLines

RedSparr0w committed:
Update Tutorial stuff

RedSparr0w committed:
Load our save object or the default save data (when starting new game), Fixup tutorial quest

RedSparr0w committed:
Remember position in questline

RedSparr0w committed:
tutorialComplete -> isTutorialCompleted

RedSparr0w committed:
add comment

RedSparr0w committed:
Tidy up

RedSparr0w committed:
Add custom reward function, firstWinReward function

RedSparr0w committed:
When defeating Hoenn Champion, start the Deoxys quest line

RedSparr0w committed:
rework quest customReward

RedSparr0w committed:
make observables

RedSparr0w committed:
Update old save data

RedSparr0w committed:
Start Deoxys quest line if player already defeated Hoenn Champion

RedSparr0w committed:
Add info icon to hover for quest description

RedSparr0w committed:
Update Deoxys quest steps

RedSparr0w committed:
Update quest descriptions

Aegyo commented on src/scripts/quests/Quests.ts:
```diff
@@ -0,0 +1,257 @@
+    }
+
+    get level(): KnockoutComputed<number> {
+        return ko.pureComputed((): number => {
```
I think all of these `get thing(): KnockoutComputed<T>` are going to be creating a new computed every time they are called. Is there some reason we don't want to just 
```
public level: KnockoutComputed<number> = ko.pureComputed(...
```
which would only make the computed once?

RedSparr0w replied:
Yeah, that makes a lot more sense

Aegyo commented on src/scripts/quests/Quests.ts:
```diff
@@ -0,0 +1,257 @@
+        return new Amount(Math.floor(250000 * Math.LOG10E * Math.log(Math.pow(notComplete, 4) + 1)), GameConstants.Currency.money);
+    }
+
+    public getCompletedQuest
```
What is this for?

RedSparr0w replied:
🤔 _Nothing, I think.._ 
Gone 💨

Aegyo reviewed (changes requested):
The tutorial questline is restarting from the beginning if you refresh the page. In develop, it is only making you restart the current quest within the tutorial questline.
If anyone gets past the key item buying quest, but leaves before finishing the tutorial, they will be stuck and have to restart (because you can't buy the key item a second time)

RedSparr0w committed:
get => computed, remove getCompletedQuest

RedSparr0w commented:
I wasn't able to reproduce the tutorial not re-loading from the current point:
![reload](https://user-images.githubusercontent.com/7288322/89720409-e80c8080-da25-11ea-83e2-ea93f62ecbec.gif)

Was there a certain point you reached where it was happening?

Aegyo commented:
I went as far as moving to route 2, defeated a couple of pokemon, waited for the console to say it had saved, and then refreshed
When it had reloaded, it had me back on the first quest
![quest](https://imgur.com/QH8UvT8.png)

RedSparr0w committed:
Fix dungeon ticket quest

If the player somehow ends up on this quest again but already has the dungeon token, it will auto complete

RedSparr0w commented:
Hmm, are you able to add a console.log or something to see what's being passed through the save when it happens?

I've updated the dungeon ticket quest to always start from `0` so if you already have it, it will auto complete, so you won't be soft locked.
But it's still not ideal that it's happening at all 🤔

Aegyo reviewed (approved):
Note to self: restart gulp if there is a new version update...

RedSparr0w committed:
Merge branch 'develop' into questLines
